### PR TITLE
fileserver: Read Etags from precomputed files

### PR DESCRIPTION
**Pull request by armadi1809**
Initial pass at closing #5734

### CODE REVIEW

**Review by jdog10141 — APPROVED** [review]

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for the great enhancement! This looks pretty good but I just have a few comments.

**Comment by armadi1809** [comment]
Thanks for taking a look at this Matt. All the feedback discussions should now be resolved.

File: modules/caddyhttp/fileserver/staticfiles.go [hunk 1]
```diff
@@ -639,6 +659,23 @@ func calculateEtag(d os.FileInfo) string {
...
 }
 
+// Finds the first corresponding etag file for a given file in the file system and return its content
+func (fsrv *FileServer) getEtagFromFile(fileSystem fs.FS, filename string) (string, error) {
+	for _, suffix := range fsrv.EtagFileExtensions {
+		etagFilename := filename + suffix
+		_, err := fs.Stat(fileSystem, etagFilename)
+		if err != nil {
+			continue
+		}
```
Comment (mholt) [inline modules/caddyhttp/fileserver/staticfiles.go:669]: Is the purpose of this to check whether we can access the file? Maybe we can eliminate this system call and instead just call `fs.ReadFile()`. Then:

```go
if errors.Is(err, fs.ErrNotExist) || errors.Is(err, fs.ErrPermission) {
    continue
}
if err != nil {
    return "", fmt.Errorf("cannot read etag from file %s: %v", etagFilename, err)
}
```

What do you think?

Comment (armadi1809) [reply modules/caddyhttp/fileserver/staticfiles.go]: Much much better. Just changed it. Thanks!

**Review by mholt — APPROVED** [review]

**Comment by mholt** [comment]
Awesome, thanks for working on this!

**Comment by hanikesn** [comment]
Nice! I wish there were docs for this. I stumbled upon this while looking in the source code to check why my custom etag headers weren't used to serve 304s.

**Comment by hanikesn** [comment]
This only seems to work for precompressed files though, any particular reason for that? I'd be quite useful for this to work on any files.

**Comment by mholt** [comment]
Docs are being updated with the new features shortly :)

**Comment by mholt** [comment]
If the etag is weak, though, the quotes won't be around the whole thing.

We could probably trim newlines if that's a common occurrence, but it's not something I've encountered in my (limited) experience.

**Comment by hanikesn** [comment]
> We could probably trim newlines if that's a common occurrence, but it's not something I've encountered in my (limited) experience.

As long as it ends up in the docs I'm happy :slightly_smiling_face: It's easy to waste time with this digging into the source code of `http.serveContent()`, but also pre-computing etags is already a quite advanced use case. With reproducible builds on the advance though it'll become more common.

**Comment by mholt** [comment]
I'd welcome a PR at least for the newline patch!

**Pull request merged**